src/NumericalAlgorithms/RootFinding/GslMultiRoot.hpp

```diff
@@ -0,0 +1,364 @@
+  /// GSL documentation for more details.
+  Hybrid,
+  /// Modification on Newton's method
+  /// that improves global convergence
+  Newton
+};
+
+/*!
+ *  \ingroup NumericalAlgorithmsGroup
+ *  \brief The different options for the convergence criterion of gsl_multiroot.
+ */
+enum class StoppingCondition {
+  /// Requires an absolute tolerance as
+  /// well as a relative tolerance.
+  /// See GSL documentation for
+  /// gsl_multiroot_test_delta.
+  TestDelta,
+  /// Requires only an absolute
+  /// tolerance. See GSL documentation
+  /// for gsl_multiroot_test_residual.
+  TestResidual
+};
+
+namespace detail {
+CREATE_IS_CALLABLE(jacobian)
+
+template <typename SolverType, typename SolverAlloc, typename SolverSet,
+          typename SolverIterate, typename SolverFree, size_t Dim,
+          typename Function>
+std::array<double, Dim> gsl_multiroot_impl(
+    Function& f, const std::array<double, Dim>& initial_guess,
+    double absolute_tolerance, size_t maximum_iterations,
+    double relative_tolerance, Verbosity verbosity,
+    double maximum_absolute_tolerance, SolverType solver_type,
+    StoppingCondition condition, SolverAlloc solver_alloc, SolverSet solver_set,
+    SolverIterate solver_iterate, SolverFree solver_free) {
+  // Supply gsl_root with the initial guess:
+  gsl_vector* const gsl_root = gsl_vector_alloc(Dim);
+  gsl_vector_set_with_std_array(gsl_root, initial_guess);
+  auto* solver = solver_alloc(solver_type, Dim);
+  solver_set(solver, &f, gsl_root);
+
+  // Take iterations:
+  if (verbosity == Verbosity::Verbose or verbosity == Verbosity::Debug) {
+    Parallel::printf("\nAttempting a root find...\n");
+  }
+  int status;
+  size_t iteration_number = 0;
+  do {
+    if (verbosity == Verbosity::Debug) {
+      print_state<Dim>(iteration_number, solver, iteration_number == 0);
+    }
+    iteration_number++;
+    status = solver_iterate(solver);
+    // Check if solver is stuck
+    if (status == GSL_ENOPROG) {
+      break;
+    }
+    if (condition == StoppingCondition::TestDelta) {
+      status = gsl_multiroot_test_delta(solver->dx, solver->x,
+                                        absolute_tolerance, relative_tolerance);
+    } else if (condition == StoppingCondition::TestResidual) {
+      // NOTE: Sometimes hybridsj works only with the test_residual condition
+      status = gsl_multiroot_test_residual(solver->f, absolute_tolerance);
+    } else {
+      ERROR(
+          "Invalid stopping condition. Has to be either TestDelta"
+          "or TestResidual.");
+    }
+  } while (status == GSL_CONTINUE and iteration_number < maximum_iterations);
+  if (verbosity == Verbosity::Verbose or verbosity == Verbosity::Debug) {
+    print_state<Dim>(iteration_number, solver, verbosity == Verbosity::Verbose);
+  }
+  bool success = (status == GSL_SUCCESS);
+  if (verbosity != Verbosity::Silent) {
+    Parallel::printf("\n");
+    if (not success) {
+      const std::string ascii_divider = std::string(80, '#');
```

Comment:
Do 70 instead of 80. The reason is sometimes if you are at exacty 80 it'll add a linewrap anyway.